subiquity/ui/views/drivers.py

```diff
@@ -0,0 +1,124 @@
+# Copyright 2021 Canonical, Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+""" Module defining the view for third-party drivers installation.
+
+"""
+import asyncio
+import logging
+from typing import List, Optional
+
+from urwid import (
+    connect_signal,
+    Text,
+    )
+
+from subiquitycore.ui.buttons import ok_btn
+from subiquitycore.ui.form import (
+    Form,
+    BooleanField,
+)
+from subiquitycore.ui.spinner import Spinner
+from subiquitycore.ui.utils import screen
+from subiquitycore.view import BaseView
+
+
+log = logging.getLogger('subiquity.ui.views.drivers')
+
+
+class DriversForm(Form):
+    """ Form that shows a checkbox to configure whether we want to install the
+    available drivers or not. """
+
+    cancel_label = _("Back")
+
+    install = BooleanField(_("Install the drivers"))
+
+
+class DriversView(BaseView):
+
+    title = _("Third-party drivers.")
+
+    form = None
+
+    def __init__(self, controller, drivers: Optional[List[str]],
+                 install: bool) -> None:
+        self.controller = controller
+
+        if drivers is None:
+            self.make_waiting(install)
+        else:
+            self.make_main(install, drivers)
+
+    def make_waiting(self, install: bool) -> None:
+        """ Change the view into a spinner and start waiting for drivers
+        asynchronously. """
+        self.spinner = Spinner(self.controller.app.aio_loop, style='dots')
+        self.spinner.start()
+        rows = [
+            Text(_("Looking for applicable third-party drivers...")),
+            Text(""),
+            self.spinner,
+            ]
+        btn = ok_btn(_("Continue"), on_press=lambda sender: self.done(False))
+        self._w = screen(rows, [btn])
+        asyncio.create_task(self._wait(install))
+
+    async def _wait(self, install: bool) -> None:
+        """ Wait until the "list" of drivers is available and change the view
+        accordingly. """
+        drivers = await self.controller._wait_drivers()
```

Comment:
We recently merged a [fix](https://github.com/canonical/subiquity/pull/1192) for an issue where the task to check if a subiquity update was available was causing answers automated install hangs.

I wonder if we need to do that for every task that is implemented on the server as a network check - that we should not leave a task around.  Once we have made the decision to continue, we should cancel that task.  We also probably want to wrap the task in a `schedule_task` call.

What do you think?

Reply:
I think this is good hygiene in general

Reply:
The problem I have with cancelling the task is that upon hitting "back" from the later screen, the user is not going to have a second chance of getting the relevant information shown.

I haven't dug enough into the root-cause of https://github.com/canonical/subiquity/pull/1192 to understand if other screens can be affected or if it's just because the "Refresh" screen is shown at two possible positions.

Any thoughts?

Reply:
I implemented this change.

The task is now cancelled on POST request to /drivers. Subsequent GET requests (if any) having wait=true will restart the task.

Unfortunately, if any task/coroutine is `await`'ing on the `_drivers_task` when it gets cancelled, an `asyncio.CancelledError` exception is raised and the server exits silently. This is not specific to the third-party drivers. I was able to reproduce this behavior in any other controller.

In the normal scenario:
* the client sends a `GET /drivers` query without `wait=true` -> the server returns immediately without any info about the drivers
* the client sends another `GET /drivers` query with `wait=true` and the server "blocks" until it can respond
* while it waits for an answer, the client can send a `POST /drivers` query to skip the third-party drivers page. This will effectively make the server raise a `asyncio.CancelledError` exception in the GET handler.

Therefore, we have to catch this exception in the GET handler. And what to return to the client when a task cancelled is another problem to solve.
I figured it would be good not to respond to the client with a normal body with "fake" data in it.
I implemented a mechanism that - upon raising a `HTTPConflict` exception from the server - raises a `ServerTaskCancelled` on the client side. The client can then handle this kind of exception using various means.

Reply:
Hm. One of the areas where asyncio is objectively terrible is cancellation, unfortunately. One annoying thing is that if there is an expectation that a task is going to be cancelled, you have to be careful to think if any await in it should have an asyncio.shield() wrapped around it. I think we're OK here on this.

I guess for the single client case, in some sense the GET(wait=True) task should be cancelled when the user clicks continue. But there could be other clients in the same case (even if that's not very likely), and clearly having the server silently exit is bad! (I'm actually surprised by that -- exceptions that reach the top level are supposed to be handled by SubiquityServer._handle_exception which puts things into an error report and sets an error state but not exit. Do you know what's going on there? Is something handling CancelledError specially?). So I think this solution is fine.

Reply:
Thanks mwhudson! What you said about the exceptions being handled by `_handle_exception` made complete sense so I went to look again at what was happening.

Turns out I misinterpreted my observations. Here's what actually happens when the server raises a `CancelledError` in a `GET` handler. 
TL;DR The server does not exit. The client actually dies and as the result sends a SIGINT to the server before exiting (at least in dry-run mode)

1. The server catches the `CancelledError` exception.
2. As the result of the exception, the server closes the socket associated with the GET query without sending a response.
3. This makes the client receive a `EPIPE` error (or something equivalent for UNIX domain sockets) that is transformed by `aiohttp` into a `aiohttp.client_exceptions.ServerDisconnectedError`.
4. The client has no handler setup for `ServerDisconnectedError` so it dies
5. Before dying, the client kills the server (as it usually does):

_generating crash report
report saved to .subiquity/var/crash/1647245854.068176031.ui.crash
**killing server 25055**_


```
report saved to .subiquity/var/crash/1647245854.068176031.ui.crash
killing server 25055
Traceback (most recent call last):
[...]
  File "/home/olivier/dev/canonical/subiquity/subiquity/ui/views/drivers.py", line 92, in _wait
    response: DriversResponse = await self.endpoint.GET(wait=True)
  File "/home/olivier/dev/canonical/subiquity/subiquity/common/api/client.py", line 45, in impl
    async with make_request(
[...]
  File "/usr/lib/python3/dist-packages/aiohttp/streams.py", line 604, in read
    await self._waiter
aiohttp.client_exceptions.ServerDisconnectedError: Server disconnected
```

thanks!

Reply:
As discussed with @dbungert, I reverted this change and will open a follow-up PR with it.